README.md
# openeo-flood-mapper (local)

## Description
This project represents a workflow retrieving flooded surfaces from intensity (backscatter) data of Sentinel-1. The
workflow is developed as part of the **interTwin** project by the **Technische Universität Wien**. As part of an 
interdisiplinary Digital Twin Engine (DTE) it is built based on the **openEO** syntax.

The current workflow uses openEO locally and provides [Python scripts](src/openeo_flood_mapper_local/) to prepare the 
input data for this purpose. Additionally, the workflow is represented and described in the dedicated 
[notebooks](notebooks/1_yeoda_dc.qmd).


CHANGELOG.rst
=========
Changelog
=========

Version 0.1.0
=============

- Intial version


src/openeo_flood_mapper_local/restructure_plia.py
import argparse
import re
from pathlib import Path

import rioxarray
import xarray as xr
from eotransform_pandas.filesystem.gather import gather_files
from eotransform_pandas.filesystem.naming.geopathfinder_conventions import yeoda_naming_convention

DATA_VERSION = 'V01R03'


def restructure_plia(root: Path, out: Path, tile_long_name: str) -> None:
    grid, tile = tile_long_name.split('_')
    grid = f'EQUI7_{grid}'
    plia_df = gather_files(root, yeoda_naming_convention, [re.compile('PLIA-TAG'),
                                                             re.compile(DATA_VERSION),
                                                             re.compile(grid),
                                                             re.compile(tile)], index='extra_field')
    out_dir = out / DATA_VERSION / grid / tile
    out_dir.mkdir(parents=True, exist_ok=True)

    for _, row in plia_df.iterrows():
        file = row['filepath']
        da = rioxarray.open_rasterio(file, mask_and_scale=True, chunks=500).squeeze()
        if row['var_name'] == "PLIA-TAG-NOBS":
            da.encoding.update({'_FillValue': -9999, 'dtype': 'int16', 'zlib': True})
        else:
            da = da / 100
            da.encoding.update({'scale_factor': 0.01, '_FillValue': -9999, 'dtype': 'int16', 'zlib': True})
        da.to_dataset(name='PLIA').to_netcdf(out_dir / f"{file.stem}.nc", mode='w')


def main():
    parser = argparse.ArgumentParser(description="Restructure plia tiffs to be loaded as local openEO dataset.")
    parser.add_argument('root', type=Path, help="root path to the yeoda file structure")
    parser.add_argument('out', type=Path, help="root path to output structure")
    parser.add_argument('tile', type=str, help='long name of tile to process, i.e. "EU020M_E051N015T3"')
    args = parser.parse_args()

    restructure_plia(args.root, args.out, args.tile)


if __name__ == '__main__':
    main()


src/openeo_flood_mapper_local/restructure_sig0.py
import argparse
import re
from pathlib import Path

import rioxarray
import xarray as xr
from eotransform_pandas.filesystem.gather import gather_files
from eotransform_pandas.filesystem.naming.geopathfinder_conventions import yeoda_naming_convention

DATA_VERSION = 'V1M1R1'


def restructure_sig0(root: Path, out: Path, tile_long_name: str) -> None:
    grid, tile = tile_long_name.split('_')
    grid = f'EQUI7_{grid}'
    sig0_df = gather_files(root, yeoda_naming_convention, [re.compile('SIG0'),
                                                             re.compile(DATA_VERSION),
                                                             re.compile(grid),
                                                             re.compile(tile)], index='datetime_1')
    out_dir = out / DATA_VERSION / grid / tile
    out_dir.mkdir(parents=True, exist_ok=True)

    for _, row in sig0_df.iterrows():
        file = row['filepath']
        da = rioxarray.open_rasterio(file, mask_and_scale=True, chunks=500).squeeze()
        da.encoding.update({'scale_factor': 0.1, '_FillValue': -9999, 'dtype': 'int16', 'zlib': True})
        da.to_dataset(name='SIG0').to_netcdf(out_dir / f"{file.stem}.nc", mode='w')


def main():
    parser = argparse.ArgumentParser(description="Restructure sig0 tiffs to be loaded as local openEO dataset.")
    parser.add_argument('root', type=Path, help="root path to the yeoda file structure")
    parser.add_argument('out', type=Path, help="root path to output structure")
    parser.add_argument('tile', type=str, help='long name of tile to process, i.e. "EU020M_E051N015T3"')
    args = parser.parse_args()

    restructure_sig0(args.root, args.out, args.tile)


if __name__ == '__main__':
    main()


src/openeo_flood_mapper_local/restructure_hparam.py
import argparse
import re
from pathlib import Path

import xarray as xr
from eotransform_pandas.filesystem.gather import gather_files
from eotransform_pandas.filesystem.naming.geopathfinder_conventions import yeoda_naming_convention

DATA_VERSION = 'V0M2R3'


def restructure_hparam(root: Path, out: Path, tile_long_name: str) -> None:
    grid, tile = tile_long_name.split('_')
    grid = f'EQUI7_{grid}'
    hparam_df = gather_files(root, yeoda_naming_convention, [re.compile('SIG0-HPAR'),
                                                             re.compile(DATA_VERSION),
                                                             re.compile(grid),
                                                             re.compile(tile)], index='extra_field')
    out_dir = out / DATA_VERSION / grid / tile
    out_dir.mkdir(parents=True, exist_ok=True)

    for orbit, grouped_df in hparam_df.groupby('extra_field'):
        da = xr.open_mfdataset(grouped_df['filepath'], concat_dim='band', combine='nested', chunks=500,
                               mask_and_scale=True)['band_data']
        var_names = [v.split('-')[-1] for v in grouped_df['var_name'].to_list()]
        da = da.assign_coords({'band': ('band', var_names)})
        ds = da.to_dataset(dim='band')
        encoding = {dv: {'scale_factor': 0.1, '_FillValue': -9999, 'dtype': 'int16', 'zlib': True}
                    for dv in ds.data_vars if dv != 'NOBS'}
        encoding['NOBS'] = {'_FillValue': -9999, 'dtype': 'int16', 'zlib': True}
        ds.to_netcdf(out_dir / f"{orbit}.nc", mode='w', encoding=encoding)


def main():
    parser = argparse.ArgumentParser(description="Restructure hparam tiffs to be loaded as local openEO dataset.")
    parser.add_argument('root', type=Path, help="root path to the yeoda file structure")
    parser.add_argument('out', type=Path, help="root path to output structure")
    parser.add_argument('tile', type=str, help='long name of tile to process, i.e. "EU020M_E051N015T3"')
    args = parser.parse_args()

    restructure_hparam(args.root, args.out, args.tile)


if __name__ == '__main__':
    main()
